File: frontend/src/debugBatch.ts
export type DebugBatchCapabilities = {
  developer_tools_enabled: boolean;
};

export type DebugBatchStartInput = {
  label: string;
  targetProjectCount: string | number;
  notes?: string;
  baselineBatchId?: string;
};

export type DebugBatchStartPayload = {
  label: string;
  target_project_count: number;
  notes?: string;
  baseline_batch_id?: string;
  frontend_build_identity: string;
};

export type DebugBatchMembership = {
  project_id: string;
  position: number;
  project_name?: string | null;
  missing: boolean;
  workflow_phase: string;
  worker_reached: boolean;
  current_working_revision_id: string | null;
  attempt_count: number;
  retry_count: number;
  provider_call_count: number;
  provider_retry_count: number;
  content_repair_count: number;
  generation_attempt_count: number;
  workflow_stage_attempt_count: number;
  user_operation_count: number;
  outcome_category: string;
  final_outcome: string;
};

export type DebugBatch = {
  id: string;
  label: string;
  notes: string | null;
  target_project_count: number;
  baseline_batch_id: string | null;
  state: "active" | "finishing" | "frozen" | "failed" | string;
  git_head: string;
  branch: string;
  migration_head: string;
  application_version: string;
  frontend_build_identity: string;
  backend_build_identity: string;
  worker_build_identity: string;
  provider: string;
  configured_default_model: string;
  stage_model_policy: Record<string, unknown>;
  actual_provider_models: Record<string, unknown>;
  prompt_versions: Record<string, unknown>;
  configuration_hash: string;
  started_at: string;
  finished_at: string | null;
  report_path: string | null;
  report_generation_state: string;
  evidence_contract_version: string;
  comparison_status: string;
  redaction_status: string;
  integrity_status: string;
  memberships: DebugBatchMembership[];
};

export type DebugBatchReport = {
  batch: DebugBatch;
  summary: Record<string, unknown>;
  report_path: string | null;
  codex_review_instruction: string | null;
};

export type DebugBatchComparison = {
  batch_id: string;
  baseline_batch_id: string;
  status: string;
  identity_match: boolean;
  mismatches: Record<string, Record<string, unknown>>;
  identity_evidence: Record<string, unknown>;
  project_comparisons: Array<Record<string, unknown>>;
};

export function normalizeDebugBatchStart(input: DebugBatchStartInput): {
  label: string;
  targetProjectCount: number;
  notes?: string;
  baselineBatchId?: string;
} {
  const label = input.label.trim();
  if (!label) {
    throw new Error("Batch name is required");
  }
  const targetProjectCount = Number(input.targetProjectCount);
  if (!Number.isInteger(targetProjectCount) || targetProjectCount < 1 || targetProjectCount > 20) {
    throw new Error("Target projects must be between 1 and 20");
  }
  const notes = input.notes?.trim() || undefined;
  const baselineBatchId = input.baselineBatchId?.trim() || undefined;
  return { label, targetProjectCount, notes, baselineBatchId };
}

export function buildDebugBatchStartPayload(
  input: DebugBatchStartInput,
  frontendBuildIdentity: string,
): DebugBatchStartPayload {
  const normalized = normalizeDebugBatchStart(input);
  return {
    label: normalized.label,
    target_project_count: normalized.targetProjectCount,
    ...(normalized.notes ? { notes: normalized.notes } : {}),
    ...(normalized.baselineBatchId ? { baseline_batch_id: normalized.baselineBatchId } : {}),
    frontend_build_identity: frontendBuildIdentity,
  };
}

const OUTCOME_LABELS: Record<string, string> = {
  not_started: "Not started",
  in_progress: "In progress",
  waiting_for_clarification: "Waiting for clarification",
  working_version_created: "Working version created",
  accepted: "Accepted",
  accepted_with_warnings: "Accepted with warnings",
  candidate_created: "Candidate created",
  blocked_before_worker: "Blocked before worker",
  blocked_after_worker: "Blocked after worker",
  interrupted: "Interrupted",
  infrastructure_failure: "Infrastructure failure",
};

export function debugBatchOutcomeLabel(value: string | null | undefined): string {
  return OUTCOME_LABELS[value ?? "not_started"] ?? "Not started";
}

export function safeFrontendDebugEvent(input: Record<string, unknown>): Record<string, unknown> {
  const allowed = [
    "event_type",
    "safe_endpoint_path",
    "project_id",
    "revision_id",
    "workflow_id",
    "visible_error_kind",
    "http_status",
    "occurred_at",
  ];
  return Object.fromEntries(allowed.filter((key) => input[key] !== undefined).map((key) => [key, input[key]]));
}

export function debugBatchProjectCount(batch: DebugBatch): number {
  return batch.memberships.length;
}

File: backend/app/schemas/debug_batch.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator


class DebugBatchStart(BaseModel):
    label: str = Field(min_length=1, max_length=200)
    target_project_count: int = Field(default=5, ge=1, le=20)
    notes: str | None = Field(default=None, max_length=12000)
    baseline_batch_id: str | None = Field(default=None, min_length=1, max_length=36)
    frontend_build_identity: str = Field(default="frontend-dev", min_length=1, max_length=160)

    @field_validator("label")
    @classmethod
    def trim_label(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("batch label cannot be blank")
        return trimmed

    @field_validator("notes")
    @classmethod
    def trim_notes(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class DebugBatchMembershipRead(BaseModel):
    project_id: str
    position: int
    project_name: str | None = None
    missing: bool = False
    workflow_phase: str = "Not started"
    worker_reached: bool = False
    current_working_revision_id: str | None = None
    attempt_count: int = 0
    retry_count: int = 0
    provider_call_count: int = 0
    provider_retry_count: int = 0
    content_repair_count: int = 0
    generation_attempt_count: int = 0
    workflow_stage_attempt_count: int = 0
    user_operation_count: int = 0
    outcome_category: str = "not_started"
    final_outcome: str = "Not started"


class DebugBatchRead(BaseModel):
    id: str
    label: str
    notes: str | None
    target_project_count: int
    baseline_batch_id: str | None
    state: str
    git_head: str
    branch: str
    migration_head: str
    application_version: str
    frontend_build_identity: str
    backend_build_identity: str
    worker_build_identity: str
    build_identities: dict[str, Any] = Field(default_factory=dict)
    identity_complete: bool = False
    provider: str
    configured_default_model: str
    stage_model_policy: dict[str, Any]
    actual_provider_models: dict[str, Any]
    prompt_versions: dict[str, Any]
    configuration_hash: str
    started_at: datetime
    finished_at: datetime | None
    report_path: str | None
    report_generation_state: str
    evidence_contract_version: str
    comparison_status: str
    redaction_status: str
    integrity_status: str
    memberships: list[DebugBatchMembershipRead] = Field(default_factory=list)


class DebugBatchFrontendEvent(BaseModel):
    event_type: str = Field(min_length=1, max_length=80)
    safe_endpoint_path: str | None = Field(default=None, max_length=240)
    project_id: str | None = Field(default=None, max_length=36)
    revision_id: str | None = Field(default=None, max_length=36)
    workflow_id: str | None = Field(default=None, max_length=36)
    visible_error_kind: str | None = Field(default=None, max_length=120)
    http_status: int | None = Field(default=None, ge=100, le=599)
    occurred_at: datetime


class DebugBatchFrontendEventsCreate(BaseModel):
    events: list[DebugBatchFrontendEvent] = Field(min_length=1, max_length=25)


class DebugBatchFrontendEventsRead(BaseModel):
    accepted_count: int


class DebugBatchSummary(BaseModel):
    batch: DebugBatchRead
    summary: dict[str, Any]
    report_path: str | None = None
    codex_review_instruction: str | None = None


class DebugBatchComparisonRead(BaseModel):
    batch_id: str
    baseline_batch_id: str
    status: str
    identity_match: bool
    mismatches: dict[str, dict[str, Any]] = Field(default_factory=dict)
    identity_evidence: dict[str, Any] = Field(default_factory=dict)
    project_comparisons: list[dict[str, Any]] = Field(default_factory=list)
